### PR TITLE
[Elastic Agent] Add upgrade CLI to initiate upgrade of Agent locally

### LINKED ISSUE #20142: [Elastic Agent] Add ability to communicate and control the running agent daemon.
To provide a rich UX for the Elastic Agent we need to provide a way to communicate and control the running Elastic Agent.

Information that needs to be able to be retrieved from the local agent:

* Running version (possible different then the executing binary, but rare)
* Current status
   * Overall agent status
   * Status of each application
* Ability to restart Elastic Agent (self-restarts self with re-exec)
* Ability to trigger self-upgrade Elastic Agent

Below documents the commands and expected outputs:

### Same version
```
$ elastic-agent version
Running Daemon: 7.9.0-SNAPSHOT (build: unknown at 0001-01-01 00:00:00 +0000 UTC)
Executable: 7.9.0-SNAPSHOT (build: unknown at 0001-01-01 00:00:00 +0000 UTC)
```

### Not running
```
$ elastic-agent version
Running Daemon: <not running>
Executable: 7.9.0-SNAPSHOT (build: unknown at 0001-01-01 00:00:00 +0000 UTC)
```

### Different versions running
```
$ elastic-agent version
WARN: The running version of Elastic Agent does not match this version.
Running Daemon: 7.9.0-SNAPSHOT (build: unknown at 0001-01-01 00:00:00 +0000 UTC)
Executable: 8.0.0-SNAPSHOT (build: unknown at 0001-01-01 00:00:00 +0000 UTC)
```

## `status`

```
$ elastic-agent status
Status: Degraded
Message: 1 application is degraded

Application           Status             Message
Filebeat              Healthy            Running
Metricbeat            Degraded           system/load is an invalid metricset
Endpoint Security     Healthy            Protected with policy {000-....}        
```

## `restart`

The different `[/]` below are showing interactive progress of what the command will report as its performing the restart action. It will then just print `Successfully restarted` over the progress.

```
$ elastic-agent restart
[/] Triggering restart
[-] Waiting to reconnect
[\] Reconnecting
[|] Connected
Successfully restarted
```

## `upgrade`

The different `[/]` below are showing interactive progress of what the command will report as its performing the restart action. It will then just print `Successfully restarted` over the progress.

### Next-version in same major

```
$ elastic-agent upgrade
[/] Triggering upgrade
[-] Upgrading to 7.9.1
[\] Reconnecting
[|] Connected
Successfully upgraded to 7.9.1
```

### Next major

```
$ elastic-agent upgrade --version 8.0.0
[/] Triggering upgrade
[-] Upgrading to 8.0.0
[\] Reconnecting
[|] Connected
Successfully upgraded to 8.0.0
```

### Specific version

```
$ elastic-agent upgrade --version 7.9.2-SNAPSHOT
[/] Triggering upgrade
[-] Upgrading to 8.9.2-SNAPSHOT
[\] Reconnecting
[|] Connected
Successfully upgraded to 7.9.2-SNAPSHOT
```

### PR DESCRIPTION
**Pull request by blakerouse**
## What does this PR do?

Adds a new `upgrade` CLI command that communicates with the running Elastic Agent to perform self-upgrading. This also allows a local upgrade to a new build using the `--source-uri file://${path}` command line flag.

## Why is it important?

To allow upgrading of the installed Elastic Agent, as well as to help with testing the upgrade process because of the allowed `file://` prefix in the `--source-uri`.

## Checklist

- [X] My code follows the style guidelines of this project
- [X] I have commented my code, particularly in hard-to-understand areas
- ~~[ ] I have made corresponding changes to the documentation~~
- ~~[ ] I have made corresponding change to the default configuration files~~
- ~~[ ] I have added tests that prove my fix is effective or that my feature works~~
- [x] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.

## How to test this PR locally

Using a properly installed Elastic Agent with `./elastic-agent install`. You can initiate an upgrade:

```
$ elastic-agent upgrade 8.0.0
```

The above will fail because there is no pushed 8.0.0, but using the same command and the added `--source-uri` upgrading can be tested.

```
$ elastic-agent upgrade 8.0.0 --source-uri file://$PWD/build/distributions
```

This requires that the branch was build with DEV on so that the *.asc files are not required `DEV=true mag package`. You can point it to the same installed version but that will result in the following error:

```
$ elastic-agent upgrade 8.0.0 --source-uri file://$PWD/build/distributions
The upgrade process of Elastic Agent is currently EXPERIMENTAL and should not be used in production
Failed trigger upgrade of daemon: upgrading to same version: /go/src/github.com/elastic/beats/x-pack/elastic-agent/pkg/agent/application/upgrade/upgrade.go[103]: unknown error
```

Adding a commit and rebuilding will result in a new version that will successfully upgrade.

```
$ elastic-agent upgrade 8.0.0 --source-uri file://$PWD/build/distributions
The upgrade process of Elastic Agent is currently EXPERIMENTAL and should not be used in production
Upgrade triggered to version 8.0.0, Elastic Agent is currently restarting
```

## Related issues

- Closes #20142

### CODE REVIEW

File: x-pack/elastic-agent/pkg/agent/application/managed_mode.go [hunk 1]
```diff
@@ -205,6 +206,7 @@ func newManaged(
 		[]context.CancelFunc{managedApplication.cancelCtxFn},
 		reexec,
 		acker)
+	uc.SetUpgrader(managedApplication.upgrader)
```
Comment (michalpristas) [inline x-pack/elastic-agent/pkg/agent/application/managed_mode.go:209]: should we allow to upgrade agent when enrolled?

**Review by michalpristas — APPROVED** [review]
Code LGTM, not sure if we should enable manual upgrade when connected to fleet, but code wise it looks ok

File: x-pack/elastic-agent/pkg/agent/application/managed_mode.go (hunk 1 above)
Comment (blakerouse) [reply x-pack/elastic-agent/pkg/agent/application/managed_mode.go:209]: I think we should allow it, but I can disable it if we think that is correct.

**Comment by blakerouse** [comment]
Switched it to not allow local upgrade when Agent is controlled by Fleet.

**Pull request merged**